opm/simulators/linalg/bda/WellContributions.cu

```diff
@@ -0,0 +1,420 @@
+            val_pointers = new unsigned int[num_wells + 1];
+            cudaMalloc((void**)&d_val_pointers, sizeof(int) * (num_wells + 1));
+            cudaCheckLastError("apply_gpu malloc failed");
+        }
+#endif
+
+        WellContributions::~WellContributions()
+        {
+#if HAVE_CUDA
+            if(gpu_mode){
+                free_gpu();
+            }else{
+                free_cpu();
+            }
+#else
+            free_cpu();
+#endif
+        }
+
+        void WellContributions::free_cpu(){
+            delete[] Cnnzs;
+            delete[] Dnnzs;
+            delete[] Bnnzs;
+            delete[] Ccols;
+            delete[] Bcols;
+            delete[] val_pointers;
+            delete[] z1;
+            delete[] z2;
+            //delete[] Mnnzs;
+        }
+
+#if HAVE_CUDA
+        void WellContributions::free_gpu(){
+            cudaFree(d_Cnnzs);
+            cudaFree(d_Dnnzs);
+            cudaFree(d_Bnnzs);
+            cudaFree(d_Ccols);
+            cudaFree(d_Bcols);
+            delete[] val_pointers;
+            cudaFree(d_val_pointers);
+            // cudaFree(d_z1);
+            // cudaFree(d_z2);
+        }
+#endif
+
+
+        void WellContributions::apply(double *x, double *y){
+#if HAVE_CUDA
+            if (gpu_mode){
+                apply_gpu(x, y);
+            }else{
+                apply_cpu(x, y);
+            }
+#else
+            apply_cpu(x, y);
+#endif
+        }
+
+        // Apply the WellContributions, similar to StandardWell::apply()
+        // y -= (C^T *(D^-1*(   B*x)))
+        void WellContributions::apply_cpu(double *x, double *y)
+        {
+#if 0
+            // Mnnzs contains a sparse matrix with a symmetric pattern
+            // Mrows would contain 'i*val_size' for every entry i, since every row has the same number of blocks
+            // Mcols are the same as Ccols, normally, there is an entry for every block, but since all rows have the same sparsity pattern, we only have to store 1 row
+            bool dbg = false;
+            for(int i = 0; i < dim*dim*val_size*val_size; ++i){
+                if(dbg)printf("Mnnzs[%d]: %.5e\n", i, Mnnzs[i]);
+            }
+            if(dbg)printf("row_size: %u, val_size: %u\n", row_size, val_size);
+            for(int r = 0; r < val_size; ++r){
+                for(int c = 0; c < val_size; ++c){
+                    int colIdx = Ccols[c];
+                    if(dbg)printf("colIdx: %d\n", colIdx);
+                    for(int i = 0; i < dim; ++i){
+                        double sum = 0.0;
+                        for(int j = 0; j < dim; ++j){
+                            sum += Mnnzs[r * dim * dim * val_size + c * dim * dim + i * dim + j] * x[colIdx * dim + j];
+                        }
+                        if(dbg)printf("sum: %f\n", sum);
+                        y[colIdx * dim + i] -= sum;
+                    }
+                }
+            }
+            if(dbg)exit(0);
+#else
```

Comment:
I had originally intended to perform `y -= (C*D*B) * x` instead of `y -= (C*(D*(B*x)))`, since that requires 1 spmv instead of 3.

Reply:
Do mind telling us what the result was and why you went the other way? Maybe we can learn something.

Reply:
I first tried creating a new BCRSMatrix in ISTLSolverEbos, to contain all `(C*D*B)` of all wells. But the sparstity pattern of the wellcontributions changes often, and creating a new BCRSMatrix is not cheap. Luckily we found out that the `--matrix-add-wellcontributions` flag already prepares a matrix with all the indices it will ever need during the simulation.
I'm not sure why I didn't go for the first route for this PR, probably because the CPU does it like this. I started the implemention for `(C*D*B)`, but other things came up, and the second version worked and is fast enough (for NORNE).
The first version should use more memory, but it only uses square blocks, and I think it is a straightforward spmv. The second version is a bit more complex.